apm-agent-core/src/main/java/co/elastic/apm/agent/bci/ElasticApmAgent.java

```diff
@@ -335,6 +346,50 @@ public boolean matches(MethodDescription target) {
             .withExceptionHandler(PRINTING);
     }
 
+    /**
+     * Validates invariants explained in {@link ElasticApmInstrumentation#indyPlugin()}
+     *
+     * @param adviceClassName the name of the advice class
+     */
+    private static void validateAdvice(String adviceClassName) {
+        TypePool pool = new TypePool.Default.WithLazyResolution(TypePool.CacheProvider.NoOp.INSTANCE, ClassFileLocator.ForClassLoader.ofSystemLoader(), TypePool.Default.ReaderMode.FAST);
```

Comment:
If using `ThreadLocal` is a bug, we don't mind failing in runtime (meaning - when you first test your plugin), as long as we shout it out and make sure what needs to be fixed.
Since you already developed an OSGi system (Barny Felix TM, the successor of Apache Felix), your custom class loader can disallow using it with `ClassNotFoundException` and a proper message 🤭

Reply:
Brilliant idea! Done.